BlueprintUI/Sources/Animation/TransitionAnimation.swift

```diff
@@ -0,0 +1,177 @@
+import UIKit
+
+
+/**
+ Defines the animation used when an element is inserted or removed during an update cycle.
+ 
+ You may provide a standard UIKit animation, or an entirely custom animation via the `animationKind` property.
+ 
+ You can control how the animation should react to other parent animations via the `performing` property.
+ */
+public struct TransitionAnimation {
+    
+    /**
+     The kind of animation:
+     - A standard UIView.animate(with...) animation.
+     - A `CustomAnimation` that can defer to CoreAnimation, CADisplayLink, etc.
+     */
+    public var animationKind : AnimationKind
+    
+    /// When the transition animation should be performed, if nested within other transition animations.
+    public var performing : PerformRule
+    
+    /// Called when the appearance animation is completed.
+    public var completion : (Bool) -> ()
+    
+    //
+    // MARK: Initialization
+    //
+    
+    public init(
+        options : AnimationOptions = .init(),
+        animations : AnimatableViewProperties,
+        performing : PerformRule = .ifNotNested,
+        completion : @escaping (Bool) -> () = { _ in }
+    )
+    {
+        self.init(
+            animationKind: .standard(options: options, properties: animations),
+            performing: performing,
+            completion: completion
+        )
+    }
+    
+    public init(
+        custom : @escaping @autoclosure () -> AnyCustomTransitionAnimation,
+        performing : PerformRule = .ifNotNested,
+        completion : @escaping (Bool) -> () = { _ in }
+    )
+    {
+        self.init(
+            animationKind: .custom(custom),
+            performing: performing,
+            completion: completion
+        )
+    }
+    
+    public init(
+        animationKind : AnimationKind,
+        performing : PerformRule = .ifNotNested,
+        completion : @escaping (Bool) -> () = { _ in }
+    ) {
+        self.animationKind = animationKind
+        self.performing = performing
+        self.completion = completion
+    }
+}
+
+
+extension TransitionAnimation {
+    
+    /// When the animation should be performed relative to other animations
+    /// further up the tree (eg, `TransitionAnimations` for parent elements).
+    public enum PerformRule {
+        /// The animation will always be performed, even if it is nested in other transition animations.
+        case always
+        
+        /// The animation will only be performed if not nested in other transition animations.
+        case ifNotNested
```

Comment:
This is actually the primary change I wanted – visually it looks super weird to transition children with every possible nested transition, versus transitioning the outermost with the children already positioned (which is what `ifNotNested` allows). Eg, if every child's appearance is an alpha fade, you end up with visually weird timing curves because you've multiplying 70% of 70% as things fade in, etc.

Fixing this was one of the primary concerns when discussing things with @bencochran, so ccing him in if he has a better example.

Reply:
Adding docs to clear this behaviour up.